sdk/protocol-plugins/src/implementation/BaseProtocolPlugin.ts

```diff
@@ -0,0 +1,66 @@
+import { Position, ChainInfo, Maybe } from '@summerfi/sdk-common/common'
+import { LendingPool, IPoolId } from '@summerfi/sdk-common/protocols'
+import { z } from 'zod'
+import { IPositionId } from '../interfaces/IPositionId'
+import { IProtocolPlugin } from '../interfaces/IProtocolPlugin'
+import { IProtocolPluginContext } from '../interfaces/IProtocolPluginContext'
+import { ActionBuilder, ActionBuildersMap } from '@summerfi/order-planner-common/builders'
+import { steps } from '@summerfi/sdk-common/simulation'
+
+/**
+ * @class Base class for all ProtocolDataPlugins. It provides the basic functionality & fields for all protocol plugins
+ *              and implements shared functionality for late dependency injection, pool schema validation
+ *              and action building
+ */
+export abstract class BaseProtocolPlugin<PoolIdType extends IPoolId>
+  implements IProtocolPlugin<PoolIdType>
+{
+  public readonly protocol: PoolIdType['protocol']['name']
+  public readonly supportedChains: ChainInfo[]
+  public _ctx: IProtocolPluginContext | undefined
+  public readonly schema: z.ZodSchema<PoolIdType>
+  public readonly StepBuilders: Partial<ActionBuildersMap>
+
+  protected constructor(
+    protocol: PoolIdType['protocol']['name'],
+    supportedChains: ChainInfo[],
+    schema: z.ZodSchema<PoolIdType>,
+    StepBuildersMap: Partial<ActionBuildersMap>,
+  ) {
+    this.protocol = protocol
+    this.supportedChains = supportedChains
+    this.schema = schema
+    this.StepBuilders = StepBuildersMap
+  }
+
+  init(ctx: IProtocolPluginContext): void {
+    this._ctx = ctx
+  }
+
+  public get ctx(): IProtocolPluginContext {
+    if (!this._ctx) {
+      throw new Error('Context (ctx) is not initialized. Please call init() with a valid context.')
+    }
+    return this._ctx
+  }
+
+  // Protocol Data Methods
+  public abstract getPool(poolId: unknown): Promise<LendingPool>
```

Comment:
I think the current agreement is not to use public as by default methods are public, but in any case it should be consistent, so please add or remove the `public` modifier to all methods